### DISCUSSION TITLE
Can you use a more recent cabal index than nixpkgs Haskell has?

### DISCUSSION THREAD

**Discussion by ParetoOptimalDev** [srid/haskell-flake#197] [Q&A]
I think maybe you could update [all-cabal-hashes](https://github.com/NixOS/nixpkgs/commit/49670257591b049f844694d95f71859c0dbe30e7) and let your current version of nixpkgs use it?

Let me explain my problems and the solutions I've tried:

Problem: I needed to use a newer version of a package than nixpkgs has
Fix: Use the hackage version via`haskellProjects.default.packages = aeson.source = "newer version";`.

Problem: The hackage index of `nixpkgs-unstable` is too old ([from 8/17](https://github.com/NixOS/nixpkgs/commit/49670257591b049f844694d95f71859c0dbe30e7#diff-6995be506a739d9ad27269ba8fdc5b62956dad4d8ac3142de9971381597fb52aR5)) to contain my package
Solution:  Use the haskell-updates nixpkgs branch with `inputs.nixpkgs = github:nixos/nixpkgs/haskell-updates`

Problem: I get an error building my Haskell project with the haskell-updates branch
Potential Solution: Replace only the all-cabal-hashes used but use the known working `nixpkgs-unstable` to build my Haskell project.

That last one is what I'm asking about in the discussion title.

**Comment by srid** [comment]
> Potential Solution: Replace only the all-cabal-hashes used but use the known working nixpkgs-unstable to build my Haskell project.

I'm not sure, but a good place to ask: https://app.element.io/#/room/#haskell:nixos.org

**Comment by roberth** [comment]
iiuc all-cabal-hashes is only referenced by the generation _process_ and not by the generated code of the package set.
So it seems feasible after a bit of `callHackage` refactoring upstream.

**Comment by MangoIV** [comment]
perhaps another solution would be tot have a `sourceDirect` attribute that doesn't invoke `callHackage` (which uses `all-cabal-hashes`) but `callHackageDirect`

**Reply by srid** [reply]
`source` is already an overloaded option (can refer to either path or version). So we should reuse it? That might be overloading it a bit too much, though.

We can mitigate this confusion using URI protocols:

```nix
{
  aeson.source = "hackage-direct:1.5.0";  # With "1.5.0" being alias to "hackage:1.5.0"
}
```

**Reply by ParetoOptimalDev** [reply]
I would be fine with either:

- This being overloaded (though I have similar reservations)
- documenting how to use a new all-cabal-hashes

For context I run into this issue when interoperating with existing hybrid cabal/nix Haskell projects where the cabal way of things is `cabal update` often.

**Comment by ParetoOptimalDev** [comment]
I've found a potential solution that is using haskell-flake, calamity :smile: 

https://github.com/simmsb/calamity-bot/blob/54e5059fcea7c4327a2a1df77eef656b62351301/flake.nix#L47

Add all-cabal-hashes input:

```
all-cabal-hashes.url = "github:commercialhaskell/all-cabal-hashes/hackage";
all-cabal-hashes.flake = false;
```

Then use in per-system:

```
          basePackages = pkgs.haskell.packages.ghc944.override {
            all-cabal-hashes = inputs.all-cabal-hashes;
          };
```

**Reply by srid** [reply]
This would be good to document.
